server/sonar-server/src/main/java/org/sonar/server/config/ws/IndexAction.java

```diff
@@ -0,0 +1,254 @@
+
+public class IndexAction implements WsAction {
+
+  private static final String SECURED_SUFFIX = ".secured";
+  private static final String LICENSE_SUFFIX = ".license";
+
+  private static final Splitter DOT_SPLITTER = Splitter.on(".").omitEmptyStrings();
+  private static final Splitter COMMA_SPLITTER = Splitter.on(",");
+  private static final String COMMA_ENCODED_VALUE = "%2C";
+
+  private static final String PARAM_KEY = "key";
+  private static final String PARAM_COMPONENT = "resource";
+
+  private final DbClient dbClient;
+  private final ComponentFinder componentFinder;
+  private final UserSession userSession;
+  private final PropertyDefinitions propertyDefinitions;
+
+  public IndexAction(DbClient dbClient, ComponentFinder componentFinder, UserSession userSession, PropertyDefinitions propertyDefinitions) {
+    this.dbClient = dbClient;
+    this.componentFinder = componentFinder;
+    this.userSession = userSession;
+    this.propertyDefinitions = propertyDefinitions;
+  }
+
+  @Override
+  public void define(WebService.NewController context) {
+    WebService.NewAction action = context.createAction("index")
+      .setDescription("This web service is deprecated, please use api/settings/values instead.")
+      .setDeprecatedSince("6.3")
+      .setResponseExample(getClass().getResource("index-example.json"))
+      .setSince("2.6")
+      .setHandler(this);
+    action.createParam(PARAM_KEY)
+      .setDescription("Setting key")
+      .setExampleValue("sonar.technicalDebt.hoursInDay");
+    action.createParam(PARAM_COMPONENT)
+      .setDescription("Component key or database id")
+      .setExampleValue(KEY_PROJECT_EXAMPLE_001);
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    JsonWriter json = response.newJsonWriter();
+    json.beginArray();
+    doHandle(json, request);
+    json.endArray();
+    json.close();
+  }
+
+  private void doHandle(JsonWriter json, Request request) {
+    DbSession dbSession = dbClient.openSession(true);
+    try {
+      Optional<ComponentDto> component = loadComponent(dbSession, request);
+      String key = request.param(PARAM_KEY);
+      List<PropertyDto> propertyDtos = loadProperties(dbSession, component, Optional.ofNullable(key));
+      new ResponseBuilder(propertyDtos).build(json);
+    } finally {
+      dbClient.closeSession(dbSession);
+    }
+  }
+
+  private Optional<ComponentDto> loadComponent(DbSession dbSession, Request request) {
+    String component = request.param(PARAM_COMPONENT);
+    if (component == null) {
+      return Optional.empty();
+    }
+    return Optional.of(loadComponent(dbSession, component));
+  }
+
+  private ComponentDto loadComponent(DbSession dbSession, String component) {
+    try {
+      Long componentId = Long.parseLong(component);
+      return componentFinder.getById(dbSession, componentId);
+    } catch (NumberFormatException e) {
+      return componentFinder.getByKey(dbSession, component);
```

Comment:
I know no worries. It's just my reaction when I see exception used for control flow, but it's because of Java in this specific case